Rohan Bhatt - Homework 6 - MSML604

Part (a) below

In [6]:
import cvxpy as cp
import numpy as np

A = np.array([[4.0, 0.9, 2.0], [1.3,  0.6, 6.0]])
b = np.array([1.2, 0.8])
n = 3
#x is the variable we're optimizing
x = cp.Variable(3, nonneg=True)  # we set nonneg=True to make sure x >= 0
#objective function x_ilog(x_i), where we're minimizing the sum of x_i l
obj_expr = cp.sum(cp.rel_entr(x, 1))

#constraints stores a list of expression trees
constraints = []
constraints += [A @ x <= b]
constraints += [cp.sum(x) == 1]

#formulating actual problem where we're minimizing the objective function with the constraints
problem = cp.Problem(cp.Minimize(obj_expr), constraints)
result = problem.solve()
print("status:", problem.status)        
print("optimal value:", problem.value)  
print("x =", x.value)


status: optimal
optimal value: -0.41407648937777874
x = [0.08766437 0.88666249 0.02567314]


We got:
* An optimal status
* An optimal value of -0.4141 which is good because we know x_ilog(x_i) < 0 for the sum of 1-3 xilog(x_o)
* The x array makes sense because the 3 values are non-negative and sum to 1 which satisfies Ax <= b

Part (b) below:

In [8]:
A = np.array([[4.0, 0.9, 2.0], [1.3,  0.6, 6.0]])
b = np.array([1.2, 0.8])

lambda1 = cp.Variable(2, nonneg=True) #>=0 in dimension 2
#building expression sum i=1 to 3 e^-a_i^t lambda
obj_expr = cp.log_sum_exp(- (A.T @ lambda1))
#full objective function:
obj = - (b @ lambda1) - obj_expr
problem = cp.Problem(cp.Maximize(obj), [])
result = problem.solve()

print("status:", problem.status)        
print("optimal value:", problem.value)  
print("lambda =", lambda1.value)


status: optimal
optimal value: -0.4140764895864858
lambda = [0.62719848 0.5282025 ]


part (c)

In [9]:
# x^* and lambda^*:
x_star = np.array([0.08766437, 0.88666249, 0.02567314])
lambda_star = np.array([0.62719848, 0.52820250])

A = np.array([
    [4.0, 0.9, 2.0],
    [1.3, 0.6, 6.0]
])

#getting partial for each i
def partial_i(i):
    return np.log(x_star[i]) + 1 + lambda_star[0]*A[0,i] + lambda_star[1]*A[1,i]
partial = [partial_i(i) for i in range(3)]
print("partial1,2,3 =", partial)


partial1,2,3 = [np.float64(1.7612174364657207), np.float64(1.7611092555880876), np.float64(1.7613019901966067)]
